src/NATS.Client.JetStream/NatsJSContext.cs

```diff
@@ -285,49 +286,90 @@ internal async ValueTask<NatsJSResponse<TResponse>> JSRequestAsync<TRequest, TRe
         CancellationToken cancellationToken = default)
         where TRequest : class
         where TResponse : class
+    {
+        var result = await TryJSRequestAsync<TRequest, TResponse>(subject, request, cancellationToken).ConfigureAwait(false);
+        if (!result.Success)
+        {
+            throw result.Error;
+        }
+
+        return result.Value;
+    }
+
+    internal async ValueTask<NatsResult<NatsJSResponse<TResponse>>> TryJSRequestAsync<TRequest, TResponse>(
+        string subject,
+        TRequest? request,
+        CancellationToken cancellationToken = default)
+        where TRequest : class
+        where TResponse : class
     {
         if (request != null)
         {
             // TODO: Can't validate using JSON serializer context at the moment.
             // Validator.ValidateObject(request, new ValidationContext(request));
         }
 
-        await using var sub = await Connection.CreateRequestSubAsync<TRequest, TResponse>(
+        await using var sub = await Connection.CreateRequestSubAsync<TRequest, JsonDocument>(
                 subject: subject,
                 data: request,
                 headers: default,
                 replyOpts: new NatsSubOpts { Timeout = Connection.Opts.RequestTimeout },
                 requestSerializer: NatsJSJsonSerializer<TRequest>.Default,
-                replySerializer: NatsJSErrorAwareJsonSerializer<TResponse>.Default,
+                replySerializer: NatsJSJsonDocumentSerializer.Default,
                 cancellationToken: cancellationToken)
             .ConfigureAwait(false);
 
         await foreach (var msg in sub.Msgs.ReadAllAsync(cancellationToken).ConfigureAwait(false))
         {
-            if (msg.Error is { } error)
+            if (msg.HasNoResponders)
             {
-                if (error.InnerException is NatsJSApiErrorException jsError)
-                {
-                    return new NatsJSResponse<TResponse>(default, jsError.Error);
-                }
-
-                throw error;
+                return new NatsNoRespondersException();
             }
 
             if (msg.Data == null)
             {
-                throw new NatsJSException("No response data received");
+                return new NatsJSException("No response data received");
             }
 
-            return new NatsJSResponse<TResponse>(msg.Data, default);
+            // We need to determine what type we're deserializing into
+            // .NET 6 new APIs to the rescue: we can read the buffer once
+            // by deserializing into a document, inspect and using the new
+            // API deserialize to the final type from the document.
+            var jsonDocument = msg.Data;
```

Comment:
There may be a concern here, due to the fact that `JsonDocument` is `IDisposable` and based on how the [Parse function for ReadOnlySequence<byte>](https://github.com/dotnet/runtime/blob/3001129ea7b0061c68176c32ba2e3520ccbb23a8/src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs#L77-L101) looks it is a case where we are likely pulling in a buffer.

Reply:
TBH I think we are fine as long as we dispose our jsondocument after getting our part, but will check if I have time this weekend.

Reply:
it's now in `using`